common.go

```diff
@@ -78,8 +93,126 @@ type PaymentMethodDetails struct {
 	AdditionalDetails AdditionalDetails `json:"additional_details,omitempty"`
 }
 
-// PaymentMethodHref wraps PaymentMethod with associated href.
-type PaymentMethodHref struct {
-	Href          string         `json:"href"`
-	PaymentMethod *PaymentMethod `json:"payment_method"`
+// This object represents the configuration of the provider that handled the transaction,
+// as defined in your PaymentsOS Control Center account.
+// Note that the object does not include your provider authentication credentials.
+type ProviderConfiguration struct {
+	ID          string      `json:"id"`
+	Name        string      `json:"name"`
+	Description string      `json:"description"`
+	Created     json.Number `json:"created"`
+	Modified    json.Number `json:"modified"`
+	ProviderID  string      `json:"provider_id"`
+	Type        string      `json:"type"`
+	AccountID   string      `json:"account_id"`
+	Href        string      `json:"href"`
+}
+
+// The line items of the order.
+type OrderLineItem struct {
+	Name      string `json:"name"`
+	ID        string `json:"id"`
+	Quantity  int    `json:"quantity"`
+	UnitPrice int64  `json:"unit_price"`
+}
+
+// Details of the order. Note that order fields required for level 2 and 3 data, must be passed separately
+// in a Create Capture request within a level_2_3 object (fields passed in the order object only are not recognized as level 2 and 3 data fields).
+type Order struct {
+	ID                string            `json:"id"`
+	AdditionalDetails AdditionalDetails `json:"additional_details,omitempty"`
+	LineItems         []OrderLineItem   `json:"line_items"`
+	TaxAmount         int64             `json:"tax_amount"`
+	TaxPercentage     decimal.Decimal   `json:"tax_percentage"`
+}
+
+type DecodedJSON map[string]interface{}
+
+func (d *DecodedJSON) UnmarshalJSON(data []byte) error {
+	var (
+		err           error
+		unquotedData  string
+		rawSet        = make(map[string]json.RawMessage)
+		citationLevel = 0
+	)
+
+	unquotedData, err = strconv.Unquote(string(data))
+	// Error can be returned if it's not quoted string
+	if err != nil {
+		// There are not quoted data, trying to unmarshal
+		err = json.Unmarshal(data, &rawSet)
+	} else {
+		for err == nil && citationLevel < maxCitationLevel {
+			unquotedData, err = strconv.Unquote(string(data))
+			citationLevel++
+		}
+		err = json.Unmarshal([]byte(unquotedData), &rawSet)
+	}
+	if err != nil {
+		return errors.Wrap(err, "can't unmarshal to map")
+	}
+	*d = make(map[string]interface{})
+	for k, v := range rawSet {
+		dValue := make(DecodedJSON)
+		err = json.Unmarshal(v, &dValue)
+		if err != nil {
+			(*d)[k] = v
```

Comment:
Am I right that this line is executed in two cases?
1. `v` is invalid json
2. `v` is not json object (but string, for example)

I think in the first case we should return an error.
And in the second case we should unmarshal `v` into `interface{}`.

What do you think?

Reply:
You understand cases correctly.
- How can we understand what is stored in the value string or JSON object?
- I disagree that we should return an error if the JSON object is invalid. We can't control what provider stores in this field. It can't be XML or something else. Also, It doesn't affect the process because it is returned from a Bank value.  Should we mark a transaction as failed if Bank returns an incorrect JSON in the field RawResponse? - I think that we shouldn't do this.

Reply:
> How can we understand what is stored in the value string or JSON object?

I think this should work.
```go
var anyType interface{}
err = json.Unmarshal(v, &anyType)
```

> I disagree that we should return an error if the JSON object is invalid. We can't control what provider stores in this field. It can't be XML or something else. Also, It doesn't affect the process because it is returned from a Bank value. Should we mark a transaction as failed if Bank returns an incorrect JSON in the field RawResponse? - I think that we shouldn't do this.

Good point. I agree that we should not fail transaction if it is not a JSON.
But I think we should guarantee that our result is valid json. For example put invalid bytes into JSON string.

Reply:
>I think this should work.

It works for digits too. Take a look at my example: https://play.golang.org/p/egyistshyyh

> But I think we should guarantee that our result is valid json

What do we do if a Bank returns XML or something else?

Reply:
> It works for digits too. Take a look at my example: https://play.golang.org/p/egyistshyyh

What's the problem with that?

> What do we do if a Bank returns XML or something else?

I realised that current code already does what I want (about invalid JSON).

Reply:
>What's the problem with that?

It's connected to your first concern.

>I think in the first case we should return an error.

We can't identify the invalid JSON is stored in value or it's a simple string.

Reply:
> It's connected to your first concern.

I know that, and I know that it works for numbers (as well as all other json types).
And I think we should use `interface{}` if we failed to unmarshal into `DecodedJSON`.

Reply:
Why do we need to unmarshal it in this case? Could you explain it? If it's not a JSON object we can stay it, isn't it?

Reply:
What if it is quoted array?
I expect to see a slice in the unmarshaled struct. Now I will see a string.

As I understand the purpose of this struct:
1. Unmarshal json object (with any types of values)
2. Unquote all possible quoted incoming data

Current implementation supports only json object with values of following types: json object or json string.

Reply:
Why are we trying to do unquote it at all? 
Jira task does not mention anything about it and the desired example from PMS keeps raw provider responses quoted as it is.

Reply:
In the Jira task, I see that we have a lot of quoted data and our task to send and as not quoted JSON. if we sent this JSON as is we would receive in result the string value, no?

Reply:
@maratori You understand the purpose correctly but with one remark,  IMHO we shouldn't unmarshal JSON with any types of values because it isn't our task. Our task to create a tree where each peak has the type map[string]interface{}. It needs because we want to marshal it as a JSON struct.

What would happen if we had the array in one of the values?
Example

```JSON
 {"c":{"b":[2,3,4.4,5]}}
```

We unmarshal it as 
```golang
zooz.DecodedJSON{
		"c": zooz.DecodedJSON{
			"b": json.RawMessage("[2,3,4,5]"),
		},
	},
```

and we can marshal it correctly. Could you give me any example when we couldn't serialize array correctly?

Reply:
ok, @maratori return your suggestion. You suggest to unmarshal it to interface{} if we can't unmarshal it to DecodedJSON, right? What we should do if we received the error on unmarshaling to interface{}?

Reply:
> You suggest to unmarshal it to interface{} if we can't unmarshal it to DecodedJSON, right?

Yes.

> What we should do if we received the error on unmarshaling to interface{}?

The same as think as now - just use raw bytes:
```go
			(*d)[k] = v
```

Reply:
I have done it